crypto/stack/stack.c

```diff
@@ -517,11 +516,49 @@ static int qsort_thunk_cmp(const void *a, const void *b, void *data)
     }
 }
 
+static int openssl_sk_partition(OPENSSL_STACK *st, int low, int high)
```

Comment:
Note this algorithm is quadratic under adversarial input, which may be a DoS concern depending on where you all use it. For BoringSSL I did an in-place heapsort, which is still pretty compact and O(N log N). You're welcome to treat it as a contribution under our CLA if you like.
https://boringssl.googlesource.com/boringssl/+/23d58423169a2d473a6028228e46f68d1e65f503/crypto/stack/stack.c#391

Reply:
This is currently only used by the STACK_OF implementation, which I suppose is public, so some users might get hit by that, but given that we used glibc's qsort previously, and didn't get any reports, I'm not overly concerned about it in the context of this PR

That said, thank you for your contribution, I would like to keep the functionality consistent for the purposes of this PR to avoid scope creep, but I'll look at doing the replacement in a separate PR

Reply:
`qsort`, despite the name, is not necessarily quicksort. It just was once upon a time. The C standard doesn't promise any particular algorithm or running time. (Unlike C++ which promises O(N log N) for `std::sort`.)

But I would expect most libcs use an O(N log N) sort. glibc appears to start with a quicksort but then transitions to a heapsort. musl uses a variant of heapsort (smoothsort). Not positive what bionic does. I think they may actually be a quicksort, but with a smarter choice of pivot than this PR. (But probably not enough to avoid a quadratic worst case.) I haven't checked Windows or macOS.

I think that means a quicksort, particularly one with a naive choice of pivot, would actually regressing DoS risk for the public API. :-/

Reply:
FWIW, if you just ignore the problem and cast the pointer into `qsort` anyway, I expect most UBSan and CFI consumers will be fine. That's what we did in BoringSSL for a while. You would only notice if libc was itself instrumented, and that's rare.

Though we did ultimately replace the `qsort` call because one of our users started testing with a statically linked, instrumented libc, which spurred finally fixing that one.

Reply:
Thats interesting, I only added that commit (after first using qsort_r), because a ubsan error was raised for me on the comparison callback.  And qsort_r is of course unusable because its not available on platforms supported here.

Reply:
Huh, interesting. Maybe GCC UBSan vs. Clang UBSan? I mean, I think it makes sense to just fix it and avoid tempting the compiler anyway. :-) Just noting another option should you all prefer it.

`qsort_r` is a disaster. Not only is it not consistently available, but there are multiple incompatible type signatures for it. And then there's a standardized `qsort_s`, but since it's part of Annex K, no one but MSVC implements it.